### DISCUSSION TITLE
Multiple `gt` tables in a `for()` renders none of the table

### DISCUSSION THREAD

**Discussion by andrie** [quarto-dev/quarto-cli#10563] [Q&A] [answered]
### Description

I am looping over a large table to print multiple extracts from the table, using `gt` and `typst`.

Although I can manually insert two `gt` tables into a single chunk, I can't get those same tables to be displayed using a loop.

I've tried various combinations of `print()` and `output: asis` with no luck.

I'm on quarto version 1.6.4

````qmd
---
title: "Multi-page gt table"
format: typst
---

# Renders correctly manually

```{r setup}
#| include: false
library(dplyr)
library(gt)
```

```{r}
#| echo: false
#| warn: false 
mtcars |> 
  head() |> 
  gt()

mtcars |> 
  tail() |>
  gt()

```
 
# Nothing renders in a loop

```{r}
#| echo: false
cyls <- unique(mtcars$cyl)
for (cyl in cyls) {
    mtcars |> 
        filter(cyl == !!cyl) |> 
        gt()
}
```

# Setting `output: asis` with `print()` doesn't help

```{r}
#| echo: false
cyls <- unique(mtcars$cyl)
for (cyl in cyls) {
    mtcars |> 
        filter(cyl == !!cyl) |> 
        gt() |> 
        print()
}
```
````

**Comment by mcanouil** [comment, marked as answer]
In the order of the document:

- ` ```{r setup} ` is the "old way". In Quarto, prefer `#| label: setup`.
- `"Nothing renders in a loop"` That's `knitr`/R behaviour. Everything inside a loop requires an explicit `print()`.
- See https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html

For "complex R object" inside for loops in `knitr` you usually need `knitr::knit_child()`/`knitr::knit_expand()`.

One answer/solution:

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Multi-page gt table"
format: typst
---

```{r}
#| label: setup
#| include: false
library(dplyr)
library(gt)
```

```{r}
#| echo: false
#| output: asis
cyls <- unique(mtcars$cyl)
for (cyl in cyls) {
  tab <- gt(filter(mtcars, cyl == !!cyl))
  cat(sep = "\n", knitr::knit_child(quiet = TRUE, text = c(
    "```{r}",
    "#| echo: false",
    "tab",
    "```"
  )))
}
```
````

</td><td>

<img width="723" alt="image" src="https://github.com/user-attachments/assets/1a92f1d3-ccb2-42ce-8f71-8538d9705343">

</td></tr>
</table>

**Reply by andrie** [reply]
Thanks very much. I haven't seen this pattern before. Time to read the docs again :-)

**Reply by cscheid** [reply]
What's going on here is that `output: asis` tells Quarto that the output of the cell will be interpreted exactly as if it had been Markdown typed manually by the user. It's the easiest way to create content programmatically.

**Reply by andrie** [reply]
Thanks, yes - I'm familiar with `output: asis` and have used it before, but I don't think I've used `knit_child` in this way.
